### PR TITLE
Add documentation

### PR DESCRIPTION
**Pull request by siuwuncheung**

### CODE REVIEW

File: lib/utils/HDFDatabase.h [hunk 1]
```diff
@@ -111,10 +111,24 @@ class HDFDatabase : public Database
...
+     *
+     * @pre !key.empty()
+     * @pre data != 0
+     * @pre nelements > 0
+     *
+     * @param[in] key The key associated with the vector of values to be
+     *                written.
+     * @param[in] data The vector of double values to be written.
+     * @param[in] nelements The number of doubles in the vector.
+     * @param[in] precision The precision of values to be written.
```
Comment (dylan-copeland) [inline lib/utils/HDFDatabase.h:126]: `precision` is actually ignored. Should we remove it?

File: lib/linalg/NNLS.h [hunk 2]
```diff
@@ -92,6 +92,10 @@ class NNLSSolver {
      */
     void normalize_constraints(Matrix& matTrans, Vector& rhs_lb, Vector& rhs_ub);
 
+
+    /**
+     * Return the number of proecessors.
```
Comment (dylan-copeland) [inline lib/linalg/NNLS.h:97]: Typo

File: lib/utils/HDFDatabase.h [hunk 3]
```diff
@@ -138,7 +152,19 @@ class HDFDatabase : public Database
...
-    virtual int getDoubleArraySize(const std::string& key);
+    /**
+     * @brief Count the number of elements in
+     * an array of doubles associated with the supplied key
+     * from the currently open HDF5 database file.
+     *
+     * @pre !key.empty()
+     *
+     * @param[in] key The key associated with the array of values to be
+     *                read.
```
Comment (dylan-copeland) [inline lib/utils/HDFDatabase.h:163]: `read.` can go on the line above.

File: lib/utils/HDFDatabase.h [hunk 4]
```diff
@@ -138,7 +152,19 @@ class HDFDatabase : public Database
         int* data,
         int nelements);
 
-    virtual int getDoubleArraySize(const std::string& key);
+    /**
+     * @brief Count the number of elements in
```
Comment (dylan-copeland) [inline lib/utils/HDFDatabase.h:156]: This can be reformatted to use up to 80 characters per line.

File: lib/algo/NonuniformDMD.h [hunk 5]
```diff
@@ -82,6 +82,24 @@ class NonuniformDMD : public DMD
...
 protected:
+    /**
+     * @brief Obtain DMD model interpolant at desired parameter point by
+     *        interpolation of DMD models from training parameter points.
+     *
+     * @param[in] parametric_dmd    The interpolant DMD model at the desired point.
+     * @param[in] parameter_points  The training parameter points.
+     * @param[in] dmds              The DMD objects associated with
+     *                              each training parameter point.
+     * @param[in] desired_point     The desired point to create a parametric DMD at.
```
Comment (dylan-copeland) [inline lib/algo/NonuniformDMD.h:93]: The desired point at which to create a parametric DMD.

File: lib/algo/DifferentialEvolution.h [hunk 6]
```diff
@@ -28,18 +28,24 @@
...
+     * @brief Constraints to be fulfilled by the variable candidates.
      */
     struct Constraints
     {
+        /**
+         * @brief Constructor.
+         *
+         * @param[in] lower           Lower bound for constraint
+         * @param[in] upper           Upper bound for constraint
+         * @param[in] isConstrained   Whether to check the contraints
```
Comment (dylan-copeland) [inline lib/algo/DifferentialEvolution.h:47]: Typo

File: lib/algo/DMD.h [hunk 7]
```diff
@@ -118,9 +141,10 @@ class DMD
      * @brief Predict state given a time. Uses the projected initial condition of the
      *        training dataset (the first column).
      *
-     * @param[in] t The time of the outputted state
+     * @param[in] t   The time of the output state
+     * @param[in] deg The deriative degree of the output state
```
Comment (dylan-copeland) [inline lib/algo/DMD.h:145]: Typo

File: lib/utils/Database.h [hunk 8]
```diff
@@ -206,10 +228,23 @@ class Database
         double* data,
         int nelements) = 0;
 
+    /**
+     * @brief Reads a sub-array of doubles associated with the supplied key
+     * from the currently open database file.
+     *
+     * @pre !key.empty()
+     * @pre data != 0 || nelements == 0
```
Comment (pranav-vempati) [inline lib/utils/Database.h:236]: Writing the precondition as `data != nullptr` would likely be more explicit.

**Review by chldkdtn — APPROVED** [review]

File: lib/utils/HDFDatabase.h (hunk 4 above)
Comment (siuwuncheung) [reply lib/utils/HDFDatabase.h:156]: It seems that astyle does not change these documentation. I reformatted it manually.

**Review by pranav-vempati — APPROVED** [review]

**Review by dylan-copeland — APPROVED** [review]

**Pull request merged**